## Tasks:

1. The World Bank's international debt data   
2. Finding the number of distinct countries     
3. Finding out the distinct debt indicators    
4. Totaling the amount of debt owed by the countries    
5. Country with the highest debt    
6. Average amount of debt across indicators    
7. The highest amount of principal repayments   
8. The most common debt indicator   
9. Other viable debt issues and conclusion       

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

## 1. The World Bank's international debt data

In [2]:
df = pd.read_csv("DebtStats//IDS_ALLCountries_Data.csv")

In [3]:
df.describe()

1970          1971          1972          1973          1974  \
count  3.499300e+04  3.584300e+04  3.657600e+04  3.736000e+04  3.791300e+04   
mean   2.418032e+08  2.615074e+08  2.960808e+08  3.708542e+08  4.697524e+08   
std    5.287998e+09  5.676280e+09  6.335208e+09  7.936191e+09  9.931066e+09   
min   -1.684299e+09 -2.095143e+09 -4.723625e+08 -2.899000e+08 -3.117914e+09   
25%    7.220000e-02  2.380550e+00  6.050350e+00  7.668257e+00  1.392980e+01   
50%    1.654475e+06  1.826000e+06  2.359205e+06  2.848000e+06  3.481873e+06   
75%    2.174353e+07  2.423075e+07  3.030630e+07  3.611921e+07  4.597607e+07   
max    5.843440e+11  6.310940e+11  7.092430e+11  8.970180e+11  1.137400e+12   

               1975          1976          1977          1978          1979  \
count  3.868700e+04  3.916500e+04  4.005600e+04  4.071400e+04  4.099100e+04   
mean   5.624011e+08  6.298892e+08  7.691486e+08  9.111953e+08  1.087913e+09   
std    1.093636e+10  1.174206e+10  1.321511e+10  1.436936e+10  1.698454e+10   
min   -1.393498e+10 -1.060234e+10 -1.485314e+10 -2.658528e+10 -1.963556e+10   
25%    2.182210e+01  4.111829e+01  1.201814e+02  3.000000e+03  5.000000e+03   
50%    4.510768e+06  5.107200e+06  6.612052e+06  7.480864e+06  9.306000e+06   
75%    5.677565e+07  6.161305e+07  7.562230e+07  9.000000e+07  1.076840e+08   
max    1.258790e+12  1.351730e+12  1.516240e+12  1.631680e+12  1.928800e+12   

       ...          2019          2020          2021          2022  \
count  ...  5.002100e+04  5.041100e+04  1.245900e+04  1.237700e+04   
mean   ...  1.522064e+10  1.526919e+10  5.242485e+09  4.701605e+09   
std    ...  2.886527e+11  2.811448e+11  3.476405e+10  2.991221e+10   
min    ... -1.338410e+11 -1.257010e+11  0.000000e+00  0.000000e+00   
25%    ...  1.163781e+00  5.140000e-02  1.343719e+07  1.388824e+07   
50%    ...  2.845400e+07  2.858300e+07  1.229288e+08  1.245334e+08   
75%    ...  6.728168e+08  6.925054e+08  9.493626e+08  9.276454e+08   
max    ...  3.101830e+13  2.987920e+13  1.065070e+12  9.303370e+11   

               2023          2024          2025          2026          2027  \
count  1.225400e+04  1.213400e+04  1.183400e+04  1.161100e+04  1.146800e+04   
mean   4.419949e+09  4.149007e+09  3.973974e+09  3.084962e+09  2.954229e+09   
std    2.752301e+10  2.418226e+10  2.332526e+10  1.706810e+10  1.613358e+10   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    1.379260e+07  1.272786e+07  1.263271e+07  1.168700e+07  1.000000e+07   
50%    1.227250e+08  1.169949e+08  1.145540e+08  9.867047e+07  9.149800e+07   
75%    8.759689e+08  8.497433e+08  7.577243e+08  6.832353e+08  6.911090e+08   
max    8.574520e+11  7.362400e+11  6.609430e+11  4.867220e+11  4.483410e+11   

               2028  
count  1.113200e+04  
mean   3.065632e+09  
std    1.663681e+10  
min    0.000000e+00  
25%    8.588643e+06  
50%    8.066908e+07  
75%    6.063123e+08  
max    4.294500e+11  

[8 rows x 59 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77389 entries, 0 to 77388
Data columns (total 65 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Country Name           77386 non-null  object 
 1   Country Code           77384 non-null  object 
 2   Counterpart-Area Name  77384 non-null  object 
 3   Counterpart-Area Code  77384 non-null  object 
 4   Series Name            77384 non-null  object 
 5   Series Code            77384 non-null  object 
 6   1970                   34993 non-null  float64
 7   1971                   35843 non-null  float64
 8   1972                   36576 non-null  float64
 9   1973                   37360 non-null  float64
 10  1974                   37913 non-null  float64
 11  1975                   38687 non-null  float64
 12  1976                   39165 non-null  float64
 13  1977                   40056 non-null  float64
 14  1978                   40714 non-null  float64
 15  19

In [5]:
df.head(2)

Country Name Country Code Counterpart-Area Name Counterpart-Area Code  \
0  Afghanistan          AFG                 World                   WLD   
1  Afghanistan          AFG                 World                   WLD   

                                         Series Name  Series Code  1970  1971  \
0  Average grace period on new external debt comm...  DT.GPA.DPPG   NaN   NaN   
1  Average grace period on new external debt comm...  DT.GPA.OFFT   NaN   NaN   

   1972  1973  ...     2019  2020  2021  2022  2023  2024  2025  2026  2027  \
0   NaN   NaN  ...  17.9041   0.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1   NaN   NaN  ...  17.9041   0.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   2028  
0   NaN  
1   NaN  

[2 rows x 65 columns]

### Let's check Null values

In [6]:
# To show upto 500 rows in output
pd.set_option('display.max_rows', 500)

df.isnull().sum()

Country Name                 3
Country Code                 5
Counterpart-Area Name        5
Counterpart-Area Code        5
Series Name                  5
Series Code                  5
1970                     42396
1971                     41546
1972                     40813
1973                     40029
1974                     39476
1975                     38702
1976                     38224
1977                     37333
1978                     36675
1979                     36398
1980                     35609
1981                     33984
1982                     33654
1983                     33346
1984                     32858
1985                     32459
1986                     32324
1987                     32721
1988                     32902
1989                     31687
1990                     31808
1991                     31910
1992                     30713
1993                     29821
1994                     29184
1995                     28851
1996    

### Replace null values (only in Years columns) with 0.

In [7]:
selected = []
for i in df.columns:
    if i[0] in ['1','2']:
        selected.append(i)

for i in selected:
    df[i] = df[i].fillna(0)

df.head(2)

Country Name Country Code Counterpart-Area Name Counterpart-Area Code  \
0  Afghanistan          AFG                 World                   WLD   
1  Afghanistan          AFG                 World                   WLD   

                                         Series Name  Series Code  1970  1971  \
0  Average grace period on new external debt comm...  DT.GPA.DPPG   0.0   0.0   
1  Average grace period on new external debt comm...  DT.GPA.OFFT   0.0   0.0   

   1972  1973  ...     2019  2020  2021  2022  2023  2024  2025  2026  2027  \
0   0.0   0.0  ...  17.9041   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1   0.0   0.0  ...  17.9041   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   

   2028  
0   0.0  
1   0.0  

[2 rows x 65 columns]

### Let's verify if all null values have been replaced

In [8]:
df.isnull().sum()

Country Name             3
Country Code             5
Counterpart-Area Name    5
Counterpart-Area Code    5
Series Name              5
Series Code              5
1970                     0
1971                     0
1972                     0
1973                     0
1974                     0
1975                     0
1976                     0
1977                     0
1978                     0
1979                     0
1980                     0
1981                     0
1982                     0
1983                     0
1984                     0
1985                     0
1986                     0
1987                     0
1988                     0
1989                     0
1990                     0
1991                     0
1992                     0
1993                     0
1994                     0
1995                     0
1996                     0
1997                     0
1998                     0
1999                     0
2000                     0
2

### Just for practice, Let's replaced missing values in Country Name with most frequent country

In [9]:
df['Country Name']=df['Country Name'].fillna(df['Country Name'].mode()[0])

In [10]:
# display rows with null values

df1 = df[df.isna().any(axis=1)]
df1

Country Name Country Code  \
77384                                        Afghanistan          NaN   
77385                                        Afghanistan          NaN   
77386                                        Afghanistan          NaN   
77387  Data from database: International Debt Statistics          NaN   
77388                           Last Updated: 10/10/2021          NaN   

      Counterpart-Area Name Counterpart-Area Code Series Name Series Code  \
77384                   NaN                   NaN         NaN         NaN   
77385                   NaN                   NaN         NaN         NaN   
77386                   NaN                   NaN         NaN         NaN   
77387                   NaN                   NaN         NaN         NaN   
77388                   NaN                   NaN         NaN         NaN   

       1970  1971  1972  1973  ...  2019  2020  2021  2022  2023  2024  2025  \
77384   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
77385   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
77386   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
77387   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
77388   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   

       2026  2027  2028  
77384   0.0   0.0   0.0  
77385   0.0   0.0   0.0  
77386   0.0   0.0   0.0  
77387   0.0   0.0   0.0  
77388   0.0   0.0   0.0  

[5 rows x 65 columns]

In [11]:
#drop last two rows because the data doesn't make sense.

df.drop(df.tail(2).index,inplace=True) 

In [12]:
# Display rows with null values again, to check. 

df1 = df[df.isna().any(axis=1)]
df1

Country Name Country Code Counterpart-Area Name Counterpart-Area Code  \
77384  Afghanistan          NaN                   NaN                   NaN   
77385  Afghanistan          NaN                   NaN                   NaN   
77386  Afghanistan          NaN                   NaN                   NaN   

      Series Name Series Code  1970  1971  1972  1973  ...  2019  2020  2021  \
77384         NaN         NaN   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   
77385         NaN         NaN   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   
77386         NaN         NaN   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   

       2022  2023  2024  2025  2026  2027  2028  
77384   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
77385   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
77386   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[3 rows x 65 columns]

### As can be seen these rows don't have any data. 

In [13]:
df.iloc[77384:,:]

Country Name Country Code Counterpart-Area Name Counterpart-Area Code  \
77384  Afghanistan          NaN                   NaN                   NaN   
77385  Afghanistan          NaN                   NaN                   NaN   
77386  Afghanistan          NaN                   NaN                   NaN   

      Series Name Series Code  1970  1971  1972  1973  ...  2019  2020  2021  \
77384         NaN         NaN   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   
77385         NaN         NaN   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   
77386         NaN         NaN   0.0   0.0   0.0   0.0  ...   0.0   0.0   0.0   

       2022  2023  2024  2025  2026  2027  2028  
77384   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
77385   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
77386   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[3 rows x 65 columns]

In [14]:
# drop rows 77384, 77385, and 77386 because they don't contain any data

df.drop([77384, 77385,77386], inplace = True)

In [15]:
# Display rows with null values

df1 = df[df.isna().any(axis=1)]
df1

Empty DataFrame
Columns: [Country Name, Country Code, Counterpart-Area Name, Counterpart-Area Code, Series Name, Series Code, 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023, 2024, 2025, 2026, 2027, 2028]
Index: []

[0 rows x 65 columns]

### Let's check for null values in the whole dataset again. It should not have any.

In [16]:
df.isnull().sum()

Country Name             0
Country Code             0
Counterpart-Area Name    0
Counterpart-Area Code    0
Series Name              0
Series Code              0
1970                     0
1971                     0
1972                     0
1973                     0
1974                     0
1975                     0
1976                     0
1977                     0
1978                     0
1979                     0
1980                     0
1981                     0
1982                     0
1983                     0
1984                     0
1985                     0
1986                     0
1987                     0
1988                     0
1989                     0
1990                     0
1991                     0
1992                     0
1993                     0
1994                     0
1995                     0
1996                     0
1997                     0
1998                     0
1999                     0
2000                     0
2

### Check for Duplicates

In [17]:
duplicateRows = df[df.duplicated()]
duplicateRows

Empty DataFrame
Columns: [Country Name, Country Code, Counterpart-Area Name, Counterpart-Area Code, Series Name, Series Code, 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023, 2024, 2025, 2026, 2027, 2028]
Index: []

[0 rows x 65 columns]

#### No duplicates found!

### Lets create a column 'Total Debt' with sum of all the values from 1970 to 2028

In [18]:
df['Total Debt']= df.iloc[:, 6:].sum(axis=1)

In [19]:
df.head(2)

Country Name Country Code Counterpart-Area Name Counterpart-Area Code  \
0  Afghanistan          AFG                 World                   WLD   
1  Afghanistan          AFG                 World                   WLD   

                                         Series Name  Series Code  1970  1971  \
0  Average grace period on new external debt comm...  DT.GPA.DPPG   0.0   0.0   
1  Average grace period on new external debt comm...  DT.GPA.OFFT   0.0   0.0   

   1972  1973  ...  2020  2021  2022  2023  2024  2025  2026  2027  2028  \
0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   

   Total Debt  
0    103.6971  
1    103.6971  

[2 rows x 66 columns]

### Let's check country name to see if there are any discrepencies...

In [20]:
for i in df['Country Name'].unique():
    print(i)

Afghanistan
Albania
Algeria
Angola
Argentina
Armenia
Azerbaijan
Bangladesh
Belarus
Belize
Benin
Bhutan
Bolivia
Bosnia and Herzegovina
Botswana
Brazil
Bulgaria
Burkina Faso
Burundi
Cabo Verde
Cambodia
Cameroon
Central African Republic
Chad
China
Colombia
Comoros
Congo, Dem. Rep.
Congo, Rep.
Costa Rica
Cote d'Ivoire
Djibouti
Dominica
Dominican Republic
East Asia & Pacific (excluding high income)
Ecuador
Egypt, Arab Rep.
El Salvador
Eritrea
Eswatini
Ethiopia
Europe & Central Asia (excluding high income)
Fiji
Gabon
Gambia, The
Georgia
Ghana
Grenada
Guatemala
Guinea
Guinea-Bissau
Guyana
Haiti
Honduras
IDA only
India
Indonesia
Iran, Islamic Rep.
Jamaica
Jordan
Kazakhstan
Kenya
Kosovo
Kyrgyz Republic
Lao PDR
Latin America & Caribbean (excluding high income)
Least developed countries: UN classification
Lebanon
Lesotho
Liberia
Low & middle income
Low income
Lower middle income
Madagascar
Malawi
Maldives
Mali
Mauritania
Mauritius
Mexico
Middle East & North Africa (excluding high income)
Middle i

#### Upon checking the country names, the following country names do not make sense:

Least developed countries: UN classification    
Low & middle income    
Low income    
Lower middle income     
Middle income    
Upper middle income    
East Asia & Pacific (excluding high income)    
Europe & Central Asia (excluding high income)    
Latin America & Caribbean (excluding high income)    
Middle East & North Africa (excluding high income)     
Sub-Saharan Africa (excluding high income)
South Asia   
IDA only        

#### Let's remove all the rows with this data

In [21]:
rows1 = ['Least developed countries: UN classification', 'Low & middle income', 'Low income', 'Lower middle income', 'Middle income', 'Upper middle income','East Asia & Pacific (excluding high income)', 'Europe & Central Asia (excluding high income)', 'Latin America & Caribbean (excluding high income)', 'Middle East & North Africa (excluding high income)', 'Sub-Saharan Africa (excluding high income)', 'South Asia', 'IDA only']

In [22]:
#define values
#values = [value1, value2, value3, ...]

#drop rows that contain any value in the list
#df = df[df.column_name.isin(values) == False]

In [23]:
df1 = df[df['Country Name'].isin(rows1) == False]

#Let's compare df1 with df to see how many rows have been removed
df1.info()
df.info()

# 7,397 rows have been removed

<class 'pandas.core.frame.DataFrame'>
Int64Index: 69987 entries, 0 to 77383
Data columns (total 66 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Country Name           69987 non-null  object 
 1   Country Code           69987 non-null  object 
 2   Counterpart-Area Name  69987 non-null  object 
 3   Counterpart-Area Code  69987 non-null  object 
 4   Series Name            69987 non-null  object 
 5   Series Code            69987 non-null  object 
 6   1970                   69987 non-null  float64
 7   1971                   69987 non-null  float64
 8   1972                   69987 non-null  float64
 9   1973                   69987 non-null  float64
 10  1974                   69987 non-null  float64
 11  1975                   69987 non-null  float64
 12  1976                   69987 non-null  float64
 13  1977                   69987 non-null  float64
 14  1978                   69987 non-null  float64
 15  19

## 2. Finding the number of distinct countries

In [24]:
lst1 = []

for i in df1['Country Name'].unique():
    print(i)
    lst1.append(i)

Afghanistan
Albania
Algeria
Angola
Argentina
Armenia
Azerbaijan
Bangladesh
Belarus
Belize
Benin
Bhutan
Bolivia
Bosnia and Herzegovina
Botswana
Brazil
Bulgaria
Burkina Faso
Burundi
Cabo Verde
Cambodia
Cameroon
Central African Republic
Chad
China
Colombia
Comoros
Congo, Dem. Rep.
Congo, Rep.
Costa Rica
Cote d'Ivoire
Djibouti
Dominica
Dominican Republic
Ecuador
Egypt, Arab Rep.
El Salvador
Eritrea
Eswatini
Ethiopia
Fiji
Gabon
Gambia, The
Georgia
Ghana
Grenada
Guatemala
Guinea
Guinea-Bissau
Guyana
Haiti
Honduras
India
Indonesia
Iran, Islamic Rep.
Jamaica
Jordan
Kazakhstan
Kenya
Kosovo
Kyrgyz Republic
Lao PDR
Lebanon
Lesotho
Liberia
Madagascar
Malawi
Maldives
Mali
Mauritania
Mauritius
Mexico
Moldova
Mongolia
Montenegro
Morocco
Mozambique
Myanmar
Nepal
Nicaragua
Niger
Nigeria
North Macedonia
Pakistan
Panama
Papua New Guinea
Paraguay
Peru
Philippines
Romania
Russian Federation
Rwanda
Samoa
Sao Tome and Principe
Senegal
Serbia
Sierra Leone
Solomon Islands
Somalia
South Africa
Sri Lanka
St. Luc

### Number of distinct countries

In [25]:
print(f'The number of distinct countries: {len(lst1)}')

The number of distinct countries: 123


## 3. Finding out the distinct debt indicators

In [26]:
df_ind = pd.read_csv('DebtStats//IDS_SeriesMetaData.csv', encoding ='latin-1')

In [27]:
df_ind.columns

Index(['Code', 'License Type', 'Indicator Name', 'Short definition',
       'Long definition', 'Source', 'Topic', 'Dataset', 'Periodicity',
       'Aggregation method', 'General comments'],
      dtype='object')

In [28]:
lst1 = df_ind['Indicator Name'].unique()
print(f"Number of distinct debt indicators: {len(lst1)}")


Number of distinct debt indicators: 563


## 4. Totaling the amount of debt owed by the countries

### Let's group all the results by country and sum the data in the total debt column.

In [29]:
df1_grouped = df1.groupby(['Country Name']).sum() 

df1_grouped

1970          1971          1972  \
Country Name                                                               
Afghanistan                     0.000000e+00  0.000000e+00  0.000000e+00   
Albania                         0.000000e+00  0.000000e+00  0.000000e+00   
Algeria                         2.944980e+10  3.600220e+10  4.506512e+10   
Angola                          2.510000e+06  1.940000e+06  2.250000e+06   
Argentina                       1.043302e+11  1.145698e+11  1.213017e+11   
Armenia                         0.000000e+00  0.000000e+00  0.000000e+00   
Azerbaijan                      0.000000e+00  0.000000e+00  0.000000e+00   
Bangladesh                      0.000000e+00  1.536000e+07  5.704432e+09   
Belarus                         0.000000e+00  0.000000e+00  0.000000e+00   
Belize                          2.654318e+08  1.169608e+08  1.287229e+08   
Benin                           1.170849e+09  1.492963e+09  1.795374e+09   
Bhutan                          2.100000e+05  1.400000e+05  1.900000e+05   
Bolivia                         1.141438e+10  1.123379e+10  1.490897e+10   
Bosnia and Herzegovina          0.000000e+00  0.000000e+00  0.000000e+00   
Botswana                        7.580250e+08  1.621573e+09  3.409652e+09   
Brazil                          1.540056e+11  1.821015e+11  2.869157e+11   
Bulgaria                        0.000000e+00  0.000000e+00  0.000000e+00   
Burkina Faso                    9.132085e+08  9.898945e+08  1.214343e+09   
Burundi                         4.480137e+08  4.319038e+08  5.173357e+08   
Cabo Verde                      0.000000e+00  0.000000e+00  0.000000e+00   
Cambodia                        7.316291e+08  1.002203e+09  5.817986e+08   
Cameroon                        4.656163e+09  5.039546e+09  7.320206e+09   
Central African Republic        6.172048e+08  8.160818e+08  1.462206e+09   
Chad                            1.216759e+09  1.699435e+09  1.203402e+09   
China                           9.260297e+10  9.980096e+10  1.136880e+11   
Colombia                        4.810211e+10  4.863563e+10  5.703952e+10   
Comoros                         2.655081e+07  2.999852e+07  4.381216e+07   
Congo, Dem. Rep.                6.417192e+09  1.262042e+10  1.751120e+10   
Congo, Rep.                     2.853064e+09  3.381580e+09  3.987958e+09   
Costa Rica                      6.295749e+09  7.365095e+09  9.081586e+09   
Cote d'Ivoire                   8.703665e+09  1.117985e+10  1.159756e+10   
Djibouti                        8.855671e+07  4.349992e+07  1.807976e+08   
Dominica                        0.000000e+00  0.000000e+00  0.000000e+00   
Dominican Republic              8.774844e+09  9.368885e+09  1.111016e+10   
Ecuador                         9.543740e+09  9.419510e+09  1.333483e+10   
Egypt, Arab Rep.                3.475203e+10  4.129029e+10  3.784976e+10   
El Salvador                     4.134127e+09  4.859259e+09  5.933135e+09   
Eritrea                         0.000000e+00  3.000000e+04  0.000000e+00   
Eswatini                        7.715896e+08  6.231898e+08  8.289605e+08   
Ethiopia                        3.670450e+09  4.776477e+09  4.907548e+09   
Fiji                            4.866023e+08  7.842096e+08  1.001764e+09   
Gabon                           2.503083e+09  3.717980e+09  6.492992e+09   
Gambia, The                     1.958668e+08  1.873819e+08  3.231780e+08   
Georgia                         0.000000e+00  0.000000e+00  0.000000e+00   
Ghana                           1.105100e+10  1.103627e+10  1.156241e+10   
Grenada                         4.076798e+08  1.505995e+08  1.187732e+08   
Guatemala                       6.277964e+09  6.146866e+09  6.631564e+09   
Guinea                          8.087382e+09  9.517077e+09  9.763727e+09   
Guinea-Bissau                   7.873369e+07  7.865016e+07  8.770293e+07   
Guyana                          2.070357e+09  3.032428e+09  3.080525e+09   
Haiti                           8.705942e+08  9.872382e+08  1.134936e+09   
Honduras    

## 5. Country with the highest debt

### Let's arrange Total Debt data in descending order to identify the countries with highest debt

In [30]:
df1_grouped.sort_values(by=['Total Debt'], inplace=True, ascending=False)

df1_grouped

1970          1971          1972  \
Country Name                                                               
China                           9.260297e+10  9.980096e+10  1.136880e+11   
Brazil                          1.540056e+11  1.821015e+11  2.869157e+11   
Mexico                          1.390345e+11  1.468469e+11  1.701992e+11   
Russian Federation              0.000000e+00  0.000000e+00  0.000000e+00   
India                           2.688730e+11  2.389938e+11  2.516748e+11   
Indonesia                       6.977949e+10  8.639049e+10  1.145330e+11   
Turkey                          6.256498e+10  6.704200e+10  7.413668e+10   
Argentina                       1.043302e+11  1.145698e+11  1.213017e+11   
Thailand                        2.047654e+10  2.009898e+10  2.324920e+10   
Venezuela, RB                   4.091592e+10  5.302058e+10  6.413752e+10   
South Africa                    2.287753e+10  2.502703e+10  2.716933e+10   
Philippines                     3.234853e+10  3.155055e+10  4.109956e+10   
Colombia                        4.810211e+10  4.863563e+10  5.703952e+10   
Egypt, Arab Rep.                3.475203e+10  4.129029e+10  3.784976e+10   
Pakistan                        8.265451e+10  9.536817e+10  7.695607e+10   
Romania                         0.000000e+00  2.382000e+09  7.057043e+09   
Vietnam                         3.453000e+08  4.054100e+08  4.598700e+08   
Nigeria                         2.375802e+10  2.343912e+10  2.875567e+10   
Kazakhstan                      0.000000e+00  0.000000e+00  0.000000e+00   
Ukraine                         0.000000e+00  0.000000e+00  0.000000e+00   
Peru                            4.370914e+10  4.499749e+10  5.274389e+10   
Morocco                         2.289923e+10  2.499729e+10  2.472937e+10   
Bangladesh                      0.000000e+00  1.536000e+07  5.704432e+09   
Algeria                         2.944980e+10  3.600220e+10  4.506512e+10   
Iran, Islamic Rep.              1.064081e+10  1.376007e+10  1.754700e+10   
Angola                          2.510000e+06  1.940000e+06  2.250000e+06   
Lebanon                         1.335081e+09  1.364591e+09  1.122214e+09   
Ecuador                         9.543740e+09  9.419510e+09  1.333483e+10   
Sri Lanka                       9.770784e+09  1.160359e+10  1.207314e+10   
Serbia                          3.500403e+10  4.664277e+10  6.357416e+10   
Bulgaria                        0.000000e+00  0.000000e+00  0.000000e+00   
Tunisia                         1.313403e+10  1.519130e+10  1.815322e+10   
Panama                          6.869185e+09  8.338634e+09  1.146675e+10   
Dominican Republic              8.774844e+09  9.368885e+09  1.111016e+10   
Kenya                           9.562589e+09  1.000600e+10  1.254828e+10   
Cote d'Ivoire                   8.703665e+09  1.117985e+10  1.159756e+10   
Ethiopia                        3.670450e+09  4.776477e+09  4.907548e+09   
Sudan                           8.886731e+09  9.254146e+09  1.112490e+10   
Belarus                         0.000000e+00  0.000000e+00  0.000000e+00   
Jordan                          3.466014e+09  4.240527e+09  5.685819e+09   
Ghana                           1.105100e+10  1.103627e+10  1.156241e+10   
Tanzania                        6.170929e+09  4.688253e+10  2.133261e+10   
Costa Rica                      6.295749e+09  7.365095e+09  9.081586e+09   
Guatemala                       6.277964e+09  6.146866e+09  6.631564e+09   
Myanmar                         3.143257e+09  4.098994e+09  5.372160e+09   
Congo, Dem. Rep.                6.417192e+09  1.262042e+10  1.751120e+10   
Cameroon                        4.656163e+09  5.039546e+09  7.320206e+09   
Zambia                          2.555432e+10  1.324692e+10  1.764282e+10   
Uzbekistan                      0.000000e+00  0.000000e+00  0.000000e+00   
Bolivia                         1.141438e+10  1.123379e+10  1.490897e+10   
Jamaica                         1.141282e+10  1.306898e+10  1.471523e+10   
El Salvador 

### Country with highest debt: China

## 6. Average amount of debt across indicators

In [31]:
df1.columns

Index(['Country Name', 'Country Code', 'Counterpart-Area Name',
       'Counterpart-Area Code', 'Series Name', 'Series Code', '1970', '1971',
       '1972', '1973', '1974', '1975', '1976', '1977', '1978', '1979', '1980',
       '1981', '1982', '1983', '1984', '1985', '1986', '1987', '1988', '1989',
       '1990', '1991', '1992', '1993', '1994', '1995', '1996', '1997', '1998',
       '1999', '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007',
       '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016',
       '2017', '2018', '2019', '2020', '2021', '2022', '2023', '2024', '2025',
       '2026', '2027', '2028', 'Total Debt'],
      dtype='object')

In [32]:
df1_gr = df1.groupby(['Series Code']).mean() 

df1_gr = df1_gr.rename(columns={'Total Debt': 'Average of debt within Series Code'})

df1_gr

#Please see the last column for the averages.

1970          1971          1972          1973  \
Series Code                                                                 
BM.GSR.TOTL.CD     8.286746e+07  1.132458e+08  1.187444e+08  1.556920e+08   
BN.CAB.XOKA.CD    -1.453902e+07 -1.708262e+07 -4.594655e+06  5.110288e+06   
BX.GRT.EXTA.CD.DT  1.286146e+07  1.455211e+07  1.927179e+07  2.615740e+07   
BX.GRT.TECH.CD.DT  9.266016e+06  9.975122e+06  1.068325e+07  1.298187e+07   
BX.GSR.TOTL.CD     3.341870e+07  3.814380e+07  6.594846e+07  9.487529e+07   
...                         ...           ...           ...           ...   
DT.UND.PRVT.CD     2.492094e+07  3.196899e+07  3.369383e+07  4.990414e+07   
FI.RES.TOTL.CD     7.716527e+07  9.545234e+07  1.466466e+08  2.189358e+08   
FI.RES.TOTL.DT.ZS  4.135652e+01  3.388399e+01  3.639961e+01  3.844227e+01   
FI.RES.TOTL.MO     4.276827e-02  6.969163e-02  1.606529e-01  1.785042e-01   
NY.GNP.MKTP.CD     3.546638e+09  3.833673e+09  4.310358e+09  5.513832e+09   

                           1974          1975          1976          1977  \
Series Code                                                                 
BM.GSR.TOTL.CD     2.815567e+08  6.198344e+08  9.632414e+08  1.536552e+09   
BN.CAB.XOKA.CD     2.196074e+07 -9.564213e+07 -3.952971e+07 -1.466027e+08   
BX.GRT.EXTA.CD.DT  4.168496e+07  5.109992e+07  4.602593e+07  4.968472e+07   
BX.GRT.TECH.CD.DT  1.388106e+07  1.670041e+07  1.644301e+07  1.778545e+07   
BX.GSR.TOTL.CD     1.982855e+08  4.020702e+08  7.914552e+08  1.189107e+09   
...                         ...           ...           ...           ...   
DT.UND.PRVT.CD     8.054155e+07  1.163648e+08  1.497349e+08  2.116554e+08   
FI.RES.TOTL.CD     3.768377e+08  3.690248e+08  4.331551e+08  5.197067e+08   
FI.RES.TOTL.DT.ZS  5.524791e+01  5.393577e+01  5.724387e+01  4.523542e+01   
FI.RES.TOTL.MO     3.860969e-01  5.331305e-01  1.076355e+00  1.523901e+00   
NY.GNP.MKTP.CD     7.030270e+09  7.768361e+09  8.386015e+09  9.413534e+09   

                           1978          1979  ...          2020  2021  2022  \
Series Code                                    ...                             
BM.GSR.TOTL.CD     1.733371e+09  2.217369e+09  ...  5.772214e+10   0.0   0.0   
BN.CAB.XOKA.CD    -2.627665e+08 -1.492261e+08  ...  1.808435e+09   0.0   0.0   
BX.GRT.EXTA.CD.DT  5.895634e+07  7.610959e+07  ...  0.000000e+00   0.0   0.0   
BX.GRT.TECH.CD.DT  2.273699e+07  2.806886e+07  ...  0.000000e+00   0.0   0.0   
BX.GSR.TOTL.CD     1.226809e+09  1.742853e+09  ...  5.656284e+10   0.0   0.0   
...                         ...           ...  ...           ...   ...   ...   
DT.UND.PRVT.CD     2.631512e+08  3.164958e+08  ...  4.737764e+08   0.0   0.0   
FI.RES.TOTL.CD     5.787978e+08  7.132198e+08  ...  5.097233e+10   0.0   0.0   
FI.RES.TOTL.DT.ZS  2.765366e+01  2.333865e+01  ...  4.710017e+01   0.0   0.0   
FI.RES.TOTL.MO     1.502070e+00  1.564637e+00  ...  4.975918e+00   0.0   0.0   
NY.GNP.MKTP.CD     1.013163e+10  1.199355e+10  ...  2.416291e+11   0.0   0.0   

                   2023  2024  2025  2026  2027  2028  \
Series Code                                             
BM.GSR.TOTL.CD      0.0   0.0   0.0   0.0   0.0   0.0   
BN.CAB.XOKA.CD      0.0   0.0   0.0   0.0   0.0   0.0   
BX.GRT.EXTA.CD.DT   0.0   0.0   0.0   0.0   0.0   0.0   
BX.GRT.TECH.CD.DT   0.0   0.0   0.0   0.0   0.0   0.0   
BX.GSR.TOTL.CD      0.0   0.0   0.0   0.0   0.0   0.0   
...                 ...   ...   ...   ...   ...   ...   
DT.UND.PRVT.CD      0.0   0.0   0.0   0.0   0.0   0.0   
FI.RES.TOTL.CD      0.0   0.0   0.0   0.0   0.0   0.0   
FI.RES.TOTL.DT.ZS   0.0   0.0   0.0   0.0   0.0   0.0   
FI.RES.TOTL.MO      0.0   0.0   0.0   0.0   0.0   0.0   
NY.GNP.MKTP.CD      0.0   0.0   0.0   0.0   0.0   0.0   

                   Average of debt within Series Code  
Series Code                                            
BM.GSR.TOTL.CD                           1.000666e+12  
BN.CAB.XOKA.CD                           3.897334e+09  
BX.GRT.EXTA.CD.D

## 7. The highest amount of principal repayments

In [56]:
# Check for rows that contains "Principal repayments" in the series name column

bool1 = df1['Series Name'].str.contains("Principal repayments")
df1_pri = df1[bool1]

df1_pri

Country Name Country Code Counterpart-Area Name Counterpart-Area Code  \
417    Afghanistan          AFG                 World                   WLD   
418    Afghanistan          AFG                 World                   WLD   
419    Afghanistan          AFG                 World                   WLD   
420    Afghanistan          AFG                 World                   WLD   
421    Afghanistan          AFG                 World                   WLD   
...            ...          ...                   ...                   ...   
77236     Zimbabwe          ZWE                 World                   WLD   
77237     Zimbabwe          ZWE                 World                   WLD   
77238     Zimbabwe          ZWE                 World                   WLD   
77239     Zimbabwe          ZWE                 World                   WLD   
77240     Zimbabwe          ZWE                 World                   WLD   

                                             Series Name     Series Code  \
417    Principal repayments on external debt, central...  DT.AMT.DECB.CD   
418    Principal repayments on external debt, general...  DT.AMT.DEGG.CD   
419    Principal repayments on external debt, long-te...  DT.AMT.DLXF.CD   
420    Principal repayments on external debt, long-te...  DT.AMT.DLTF.CD   
421    Principal repayments on external debt, other p...  DT.AMT.DOPS.CD   
...                                                  ...             ...   
77236  Principal repayments on external debt, other p...  DT.AMT.DOPS.CD   
77237  Principal repayments on external debt, private...  DT.AMT.PRPG.CD   
77238  Principal repayments on external debt, private...  DT.AMT.DPNG.CD   
77239  Principal repayments on external debt, public ...  DT.AMT.DPPG.CD   
77240  Principal repayments on external debt, public ...  DT.AMT.DEPS.CD   

            1970       1971       1972        1973  ...         2020  \
417          0.0        0.0        0.0         0.0  ...          0.0   
418          0.0        0.0        0.0         0.0  ...   21632557.5   
419          0.0        0.0        0.0         0.0  ...   26424557.5   
420          0.0        0.0        0.0         0.0  ...   35570043.4   
421          0.0        0.0        0.0         0.0  ...          0.0   
...          ...        ...        ...         ...  ...          ...   
77236  3119900.0  2784658.4  3807945.3   5651619.4  ...          0.0   
77237        0.0        0.0        0.0         0.0  ...          0.0   
77238        0.0        0.0        0.0         0.0  ...  909550000.0   
77239  4839900.0  4759154.2  5611095.5  10573957.7  ...   20874378.4   
77240  4839900.0  4759154.2  5611095.5  10573957.7  ...   20874378.4   

               2021         2022         2023         2024         2025  \
417    0.000000e+00          0.0          0.0          0.0          0.0   
418    8.911477e+07   90229050.9   90229050.9   97215018.5   95389705.7   
419    9.390676e+07   93601063.1   93246067.1   98989998.2   96809689.5   
420    0.000000e+00          0.0          0.0          0.0          0.0   
421    0.000000e+00          0.0          0.0          0.0          0.0   
...             ...          ...          ...          ...          ...   
77236  5.357498e+06    5357498.5    4869376.4    2933919.3     998462.1   
77237  0.000000e+00          0.0          0.0          0.0          0.0   
77238  1.032224e+09  726379880.0  649918840.0  382305200.0  305844160.0   
77239  1.328941e+08  130946083.6  118502605.9  116529303.8  158823253.8   
77240  0.000000e+00          0.0          0.0          0.0          0.0   

              2026         2027         2028    Total Debt  
417            0.0          0.0          0.0  0.000000e+00  
418     94832562.8   94832562.8   94832562.8  9.111661e+08  
419     95897550.6   95719603.3   95542554.7  9.580621e+08  
420            0.0          0.0          0.0  3.275733e+08  
421            0.0          0.0          0.0  0.000000e+00  
...         

In [57]:
# Group the countries and sum the payments. Then sort the sum in descending order.b

df1_gr = df1_pri.groupby(['Country Name']).sum()

df1_gr.sort_values(by=['Total Debt'], inplace=True, ascending=False)

df1_gr = df1_gr.rename(columns={'Total Debt': 'Total Repayments'})

df1_gr


1970          1971          1972  \
Country Name                                                               
China                           0.000000e+00  0.000000e+00  0.000000e+00   
Brazil                          1.973479e+09  2.372133e+09  3.148721e+09   
Russian Federation              0.000000e+00  0.000000e+00  0.000000e+00   
Mexico                          4.029309e+09  4.171397e+09  4.750649e+09   
Turkey                          5.522628e+08  5.793990e+08  8.761955e+08   
India                           3.716908e+09  1.707084e+09  1.917459e+09   
Indonesia                       4.804184e+08  6.554168e+08  8.526644e+08   
Kazakhstan                      0.000000e+00  0.000000e+00  0.000000e+00   
Argentina                       3.010517e+09  2.782781e+09  2.936633e+09   
Thailand                        4.364098e+08  5.139614e+08  4.639006e+08   
Romania                         0.000000e+00  0.000000e+00  0.000000e+00   
Colombia                        7.111050e+08  7.366236e+08  8.345104e+08   
Ukraine                         0.000000e+00  0.000000e+00  0.000000e+00   
Venezuela, RB                   3.017692e+08  5.113932e+08  8.451983e+08   
Philippines                     9.364375e+08  9.043871e+08  1.000798e+09   
South Africa                    0.000000e+00  0.000000e+00  0.000000e+00   
Lebanon                         1.183823e+07  3.508886e+07  3.474798e+07   
Algeria                         1.728320e+08  2.631482e+08  6.921402e+08   
Angola                          0.000000e+00  0.000000e+00  0.000000e+00   
Pakistan                        7.363953e+08  5.529571e+08  5.397059e+08   
Egypt, Arab Rep.                1.132550e+09  9.444134e+08  1.425760e+09   
Vietnam                         0.000000e+00  0.000000e+00  0.000000e+00   
Morocco                         2.082331e+08  3.566179e+08  4.095985e+08   
Peru                            1.241936e+09  1.342273e+09  1.439017e+09   
Serbia                          1.475310e+09  1.925901e+09  2.592228e+09   
Ecuador                         1.326658e+08  1.721531e+08  1.826709e+08   
Bulgaria                        0.000000e+00  0.000000e+00  0.000000e+00   
Nigeria                         2.785855e+08  2.642684e+08  2.647337e+08   
Tunisia                         2.367907e+08  2.564470e+08  3.680274e+08   
Sri Lanka                       1.736688e+08  1.719243e+08  2.057352e+08   
Iran, Islamic Rep.              0.000000e+00  0.000000e+00  0.000000e+00   
Panama                          1.197750e+08  1.447533e+08  1.774885e+08   
Belarus                         0.000000e+00  0.000000e+00  0.000000e+00   
Bangladesh                      0.000000e+00  0.000000e+00  0.000000e+00   
Dominican Republic              1.039300e+08  1.328300e+08  1.272319e+08   
Costa Rica                      1.767490e+08  1.799175e+08  2.085649e+08   
Kenya                           1.246789e+08  1.386401e+08  1.052036e+08   
Jordan                          1.360018e+07  2.410024e+07  3.507029e+07   
Mauritius                       1.102350e+07  1.426566e+07  7.847143e+06   
Cote d'Ivoire                   1.483575e+08  1.540325e+08  2.047881e+08   
Mongolia                        0.000000e+00  0.000000e+00  0.000000e+00   
Guatemala                       1.216020e+08  1.141440e+08  1.992221e+08   
Uzbekistan                      0.000000e+00  0.000000e+00  0.000000e+00   
Jamaica                         5.226050e+08  5.449443e+08  5.778900e+08   
El Salvador                     8.724000e+07  1.218109e+08  1.103423e+08   
Papua New Guinea                6.127840e+07  1.298383e+08  2.390564e+08   
Zambia                          1.973002e+08  2.431968e+08  4.132747e+08   
Ethiopia                        7.335740e+07  6.489501e+07  6.550722e+07   
Ghana                           9.545385e+07  9.416629e+07  5.526238e+07   
Azerbaijan                      0.000000e+00  0.000000e+00  0.000000e+00   
Paraguay                        4.231971e+07  4.857974e+07  5.765193e+07   
Turkmenistan

#### Highest Payment Repayments was made by China and the amount is 6.283808e+12

## 8. The most common debt indicator

In [33]:
df_ind.columns

Index(['Code', 'License Type', 'Indicator Name', 'Short definition',
       'Long definition', 'Source', 'Topic', 'Dataset', 'Periodicity',
       'Aggregation method', 'General comments'],
      dtype='object')

In [63]:
df_ind['Indicator Name'].value_counts()

Average grace period on new external debt commitments (years)    1
PPG, multilateral concessional (DOD, current US$)                1
PPG, multilateral (DIS, current US$)                             1
PPG, multilateral (DOD, current US$)                             1
PPG, multilateral (INT, current US$)                             1
                                                                ..
GG, official creditors (TDS, current US$)                        1
GG, official creditors (NTR, current US$)                        1
GG, official creditors (NFL, current US$)                        1
GG, official creditors (INT, current US$)                        1
Use of IMF credit, SDR allocations (DOD, current US$)            1
Name: Indicator Name, Length: 562, dtype: int64

In [36]:
df_ind.Code.value_counts()

DT.GPA.DPPG          1
DT.DIS.MLTC.CD       1
DT.AMT.MLAT.CD       1
DT.DIS.MLAT.CD       1
DT.DOD.MLAT.CD       1
                    ..
DT.TDS.OFFT.GG.CD    1
DT.NTR.OFFT.GG.CD    1
DT.NFL.OFFT.GG.CD    1
DT.INT.OFFT.GG.CD    1
DT.DOD.DSDR.CD       1
Name: Code, Length: 569, dtype: int64

## 9. Conclusions

1. The dataset contains data on 123 countries with 563 debt indicators      
2. Top 3 countries with highest debt are China, Brazil, and Mexico        
3. Top 3 countries with highest principle repayment are China, Brazil, and Russian Federation.        